packages/zoe/src/contractFacet.js

```diff
@@ -0,0 +1,470 @@
+        ...currentAllocation,
+        ...newAllocation,
+      });
+
+      assert(
+        isOfferSafe(getAmountMathForBrand, proposal, reallocation),
+        details`The reallocation was not offer safe`,
+      );
+      return reallocation;
+    };
+
+    // Make the reallocation and test for offer safety by comparing the
+    // reallocation to the original proposal.
+    const reallocations = offerHandles.map((offerHandle, i) =>
+      makeOfferSafeReallocation(offerHandle, newAllocations[i]),
+    );
+
+    // 2. Ensure that rights are conserved overall.
+    const flattened = arr => [].concat(...arr);
+    const flattenAllocations = allocations =>
+      flattened(allocations.map(allocation => Object.values(allocation)));
+
+    const currentAllocations = offerTable
+      .getOffers(offerHandles)
+      .map(({ currentAllocation }) => currentAllocation);
+    const previousAmounts = flattenAllocations(currentAllocations);
+    const newAmounts = flattenAllocations(reallocations);
+
+    assert(
+      areRightsConserved(getAmountMathForBrand, previousAmounts, newAmounts),
+      details`Rights are not conserved in the proposed reallocation`,
+    );
+
+    // 3. Save the reallocations.
+    offerTable.updateAmounts(offerHandles, reallocations);
+    E(zoeForZcf).updateAmounts(offerHandles, reallocations);
+  };
+
+  const addNewIssuer = (issuerP, keyword) =>
+    issuerTable.getPromiseForIssuerRecord(issuerP).then(issuerRecord => {
+      E(zoeForZcf).addNewIssuer(issuerP, keyword);
+      assertKeywordName(keyword);
+      assert(
+        !getKeywords(instanceRecord.issuerKeywordRecord).includes(keyword),
+        details`keyword ${keyword} must be unique`,
+      );
+      const newIssuerKeywordRecord = {
+        ...instanceRecord.issuerKeywordRecord,
+        [keyword]: issuerRecord.issuer,
+      };
+      const newBrandKeywordRecord = {
+        ...instanceRecord.brandKeywordRecord,
+        [keyword]: issuerRecord.brand,
+      };
+      instanceRecord.brandKeywordRecord = newBrandKeywordRecord;
+      instanceRecord.issuerKeywordRecord = newIssuerKeywordRecord;
+
+      return removePurse(issuerRecord);
+    });
+
+  function completeOffers(offerHandlesToDrop) {
+    assertOffersAreActive(offerHandlesToDrop);
+    offerTable.deleteOffers(offerHandlesToDrop);
+    return E(zoeForZcf).completeOffers(offerHandlesToDrop);
+  }
+
+  /** @type {ContractFacet} */
+  const makeContractFacet = zoeService =>
+    harden({
+      reallocate,
+      addNewIssuer,
+      complete: completeOffers,
+      makeInvitation: (offerHook, inviteDesc, options = harden({})) => {
+        const inviteCallback = harden({ invoke: offerHook });
+        return E(zoeForZcf).makeInvitation(inviteCallback, inviteDesc, options);
+      },
+      initPublicAPI: newPublicAPI => {
+        assert(
+          publicAPI === undefined,
+          details`the publicAPI has already been initialized`,
+        );
+        publicAPI = newPublicAPI;
```

Comment:
Hmm, so the publicAPI on the instanceRecord in ZCF is not a promise, but the publicAPI on the instanceRecord in Zoe will be a promise? I think we should make it the same in both places.

Reply:
This publicAPI is only used to verify that we haven't previous set a value. The publicAPI in the instanceRecord we return from zcf is the promise supplied to startContract in instanceData.

Reply:
It would probably be clearer to use `let publicAPIIsSet = false`, otherwise it's possible to call `initPublicAPI(undefined); initPublicAPI(really)` and not get the assertion failure and also the publicAPI promise is resolved to undefined.